aws/amazon-chime-sdk-js
# Fix infinite loop when passing in MediaDeviceInfo when choosing Audio Input

**Issue**
If you pass in MediaDeviceInfo object into `chooseAudioInputDevice`, it will cause infinite loop. 
After requesting a device from getUserMedia, `chooseAudioInputDevice` calls `restartLocalAudio` which in turns call `chooseInputIntrinsicDevice` again and it relies on the [existing device check](https://github.com/aws/amazon-chime-sdk-js/blob/v2.6.0/src/devicecontroller/DefaultDeviceController.ts#L987) to quit early. If this check returns false, we will try to select the device again.

The root cause issue here is that the logic for [calculating media track constraints](https://github.com/aws/amazon-chime-sdk-js/blob/master/src/devicecontroller/DefaultDeviceController.ts#L1221) will return the constraints without any device info if the input is MediaDeviceInfo. Without the device info (deviceId and groupId), getUserMedia might not return the device requested.

In 2.5.0 and earlier, there is actually a bug when selecting input device. The code flow would be in Chrome assume there are 2 devices (device 1 - default, device 2)
```
chooseAudioInputDevice(deviceInfo for device 2)
->  chooseInputIntrinsicDevice(deviceInfo for device 2)
->    getUserMedia({ constraints with no deviceInfo}) = stream from default device 1
->    set activeDevices groupId from input which is device2's groupId <- incorrect
->    restartLocalAudio
->      chooseInputIntrinsicDevice(existing constraint with no deviceInfo)
->      hasSameGroupId -> returns false since the groupId recorded does not match the constraint which is ''
->      getUserMedia({ constraints with no deviceInfo}) = stream from default device 1
->      set activeDevices groupId from input which is now ''
->      restartLocalAudio
->      chooseInputIntrinsicDevice(existing constraint with no deviceInfo)
->      hasSameGroupId -> returns true since the groupId recorded is now also ''
```
End result: device1 is selected instead of device 2.

In 2.6.0, we updated the logic to actually get the groupId from the acquired stream (in PR #1078) which together with the existing constraint bug causes the infinite loop.
```
chooseAudioInputDevice(deviceInfo for device 2)
->  chooseInputIntrinsicDevice(deviceInfo for device 2)
->    getUserMedia({ constraints with no deviceInfo}) = stream from default device 1
->    set activeDevices groupId from acquired stream which is default device 1's groupId
->    restartLocalAudio
->      chooseInputIntrinsicDevice(existing constraint with no deviceInfo)
->      hasSameGroupId -> returns false since the groupId recorded does not match the constraint which is ''
->      getUserMedia({ constraints with no deviceInfo}) = stream from default device 1
->      set activeDevices groupId from acquired stream which is default device 1's groupId
->      restartLocalAudio
->      chooseInputIntrinsicDevice(existing constraint with no deviceInfo)
->      hasSameGroupId -> returns false since the groupId recorded does not match the constraint which is ''
->      getUserMedia({ constraints with no deviceInfo}) = stream from default device 1
->      <the loop continue>
```
**Description of changes:**
Fixing the constraint bug if MediaDeviceInfo is passed to chooseAudioInputDevice.
Although MediaDeviceInfo is not a valid input to chooseAudioInputDevice, it works in previous version so we try to keep the same behavior. 
**Testing**

1. Have you successfully run `npm run build:release` locally? Yes
2. How did you test these changes? Manually run demo app and call chooseAudioInput with mediaDeviceInfo
3. Can these changes be tested using the demo application? If yes, which demo application can be used to test it? Yes
1. Open console and call `devices = await app.audioVideo.listAudioInputDevices();`
2. Call await app.audioVideo.listAudioInputDevices(devices[2]);
3. Make sure that the call returns.
4. Do you add, modify, or delete public API definitions? If yes, has that been reviewed and approved? No
5. Do you change the wire protocol, e.g. the request method? If yes, has that been reviewed and approved? No


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
